jni/src/faiss_index_service.cpp

```diff
@@ -57,6 +57,104 @@ void SetExtraParameters(knn_jni::JNIUtilInterface * jniUtil, JNIEnv *env,
 
 IndexService::IndexService(std::unique_ptr<FaissMethods> faissMethods) : faissMethods(std::move(faissMethods)) {}
 
+jlong IndexService::initIndex(
+        knn_jni::JNIUtilInterface * jniUtil,
+        JNIEnv * env,
+        faiss::MetricType metric,
+        std::string indexDescription,
+        int dim,
+        int numVectors,
+        int threadCount,
+        std::unordered_map<std::string, jobject> parameters
+    ) {
+    // Removed the check for number of vectors
+    // Don't use unique_ptr here since we want to access the index in the future.
+    faiss::Index * indexWriter(faissMethods->indexFactory(dim, indexDescription.c_str(), metric));
+
+    // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
+    if(threadCount != 0) {
+        omp_set_num_threads(threadCount);
+    }
+
+    // Add extra parameters that cant be configured with the index factory
+    SetExtraParameters<faiss::Index, faiss::IndexIVF, faiss::IndexHNSW>(jniUtil, env, parameters, indexWriter);
+
+    // Check that the index does not need to be trained
+    if(!indexWriter->is_trained) {
+        throw std::runtime_error("Index is not trained");
+    }
+
+    // Add vectors
+    faiss::IndexIDMap * idMap(faissMethods->indexIdMap(indexWriter));
+
+    // Check to see if the current index is HNSW
+    faiss::IndexHNSW * hnsw = dynamic_cast<faiss::IndexHNSW *>(idMap->index);
+
+    if(hnsw != NULL) {
+        // Check to see if the HNSW storage is IndexFlat
+        faiss::IndexFlat * storage = dynamic_cast<faiss::IndexFlat *>(hnsw->storage);
+        if(storage != NULL) {
+            // Allocate enough memory for all of the vectors we plan on inserting
+            // We do this to avoid unnecessary memory allocations during insert
+            storage->codes.reserve(dim * numVectors * 4);
+        }
+    }
+
+    return (jlong)idMap;
+}
+
+void IndexService::insertToIndex(
```

Comment:
Same comment as above on making more error handling friendly , releasing memory when exception happned.

Reply:
I am now handling exceptions and freeing memory in the java layer. Resolving for now.